### PR TITLE
Fixed #28188 - fixed model field pickling.

### PR DESCRIPTION
**Pull request by adamalton**
Regression in d2a26c1a90e837777dabdf3d67ceec4d2a70fb86

https://code.djangoproject.com/ticket/28188

### CODE REVIEW

File: django/utils/functional.py [hunk 1]
```diff
@@ -32,6 +32,8 @@ def __init__(self, func, name=None):
     def __get__(self, instance, cls=None):
         if instance is None:
             return self
+        # This calls the function and then assigns the returned value in place
```
Comment (timgraham) [inline django/utils/functional.py]: I wouldn't put this mostly unrelated comment in the PR.

File: tests/model_fields/tests.py [hunk 2]
```diff
@@ -76,6 +77,13 @@ def test_field_ordering(self):
         self.assertIsNotNone(f1)
         self.assertNotIn(f2, (None, 1, ''))
 
+    def test_field_instance_is_pickleable(self):
```
Comment (timgraham) [inline tests/model_fields/tests.py]: `picklable`

File: tests/model_fields/tests.py [hunk 3]
```diff
@@ -76,6 +77,13 @@ def test_field_ordering(self):
...
+        """Field instances can be pickled."""
+        field = models.Field(max_length=100, default="a string")
+        # Access the default value, this is where sneaky lambdas like to hide
```
Comment (timgraham) [inline tests/model_fields/tests.py]: `# Field is picklable with this cached property populated (#28188).`

File: tests/model_fields/tests.py [hunk 4]
```diff
@@ -76,6 +77,13 @@ def test_field_ordering(self):
         self.assertIsNotNone(f1)
         self.assertNotIn(f2, (None, 1, ''))
 
+    def test_field_instance_is_pickleable(self):
+        """Field instances can be pickled."""
+        field = models.Field(max_length=100, default="a string")
```
Comment (timgraham) [inline tests/model_fields/tests.py]: use single quotes

File: django/db/models/fields/__init__.py [hunk 5]
```diff
@@ -775,17 +771,21 @@ def get_default(self):
         """
         Returns the default value for this field.
         """
-        return self._get_default()
+        # As much logic as possible is cached inside the _get_default method,
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: A tentative tweaking of this comment might be:
```
# As much logic as possible is cached inside _get_default(), but to
# avoid requiring _get_default() to return unpicklable lambdas, allow
# the return value to be noncallable.
```

File: tests/model_fields/tests.py (hunk 3 above)
Comment (adamchainz) [inline tests/model_fields/tests.py]: sneaky indeed!

**Comment by timgraham** [comment]
I implemented Adam J's suggested approach in https://github.com/django/django/pull/8489.

**Comment by adamalton** [comment]
Thanks for the feedback @timgraham and @adamchainz.  Even though it wasn't merged, the comments were still useful.

**Comment by adamchainz** [comment]
@adamalton what is your use case for pickling fields?

**Comment by adamalton** [comment]
We're pickling fields in [Djangae's migrations implementation](https://github.com/potatolondon/djangae/tree/master/djangae/db/migrations).  Essentially we're pickling the migration operations in order to run them in background tasks.

I would normally have tried to just work around it, but I saw that there had been other Django tickets fixed in the past for when things weren't pickle-able which led me to believe that making sure things can be pickled is probably the intended approach, hence the ticket & PR.

**Comment by adamchainz** [comment]
Ah right, interesting approach.

I don't think there's any guarantee that model fields can be picklable and nor was it necessary intended. As I wrote on Tim's follow-up PR, users could easily make a field unpicklable by passing an unpicklable validator or any number of other things, and there are no checks for that. Perhaps worth adding some system checks to Djangae.

**Comment by adamalton** [comment]
Ah yes, that's a good point.  I'll add a check into Djangae.  Thanks.

**Pull request closed without merging**